### PR TITLE
[https://nvbugs/5669671][fix] Support GuidedDecoder with sharded logits

### LINKED ISSUE #7878: [Bug]: Guided Decoding with MTP > 1, draft model produces tokens not accepted by matcher (xgrammar)
### System Info

- TensorRT-LLM commit 77657a1c1204ba4ae8eecfd0937b1cbfd78cc9b9

### Who can help?

@laikhtewari 

### Information

- [ ] The official example scripts
- [ ] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [ ] My own task or dataset (give details below)

### Reproduction

Steps to reproduce:

Run deepseek model with mtp > 1, enable guided decoding, enable debug log.
e.g. 
trtllm-serve /data/weights/pytrtllm-deepseek-ai--DeepSeek-R1-Turbo/001 --tp_size=8 --ep_size=8 --host=0.0.0.0 --port=8000 --backend=pytorch --max_num_tokens=32768 --max_seq_len 40960 --max_batch_size=8 --guided_decoding_backend=xgrammar --extra_llm_api_options='{"speculative_config":{"decoding_type":"MTP","num_nextn_predict_layers":3}}'

### Expected behavior

No errors.

### actual behavior

This debug log shows up, which indicates the bitmask is not applied correctly to draft model.
`Draft request 10 at slot 2 failed to accept last new token: 21394.`

### additional notes

This seems to only happen after the first draft output. So does not show up with MTP=1

### Before submitting a new issue...

- [x] Make sure you already searched for relevant issues, and checked the [documentation](https://nvidia.github.io/TensorRT-LLM/) and [examples](https://github.com/NVIDIA/TensorRT-LLM/tree/main/examples) for answers to frequently asked questions.

### PR DESCRIPTION
**Pull request by syuoni**
## Description

This PR supports GuidedDecoder with sharded logits, which resolves https://github.com/NVIDIA/TensorRT-LLM/issues/7878

------

When MTP is used with tensor parallelism:
* For the main model LM head, the logits are gathered immediately after the GEMM, so the logits passed to GuidedDecoder is complete.
* For the draft model LM head, the logits are NOT gathered, resulting in sharded logits passed to  GuidedDecoder.

This is the root cause of https://github.com/NVIDIA/TensorRT-LLM/issues/7878.

## PR Checklist

Please review the following before submitting your PR:
- PR description clearly explains what and why. If using CodeRabbit's summary, please make sure it makes sense.
- PR Follows [TRT-LLM CODING GUIDELINES](https://github.com/NVIDIA/TensorRT-LLM/blob/main/CODING_GUIDELINES.md) to the best of your knowledge.
- Test cases are provided for new code paths (see [test instructions](https://github.com/NVIDIA/TensorRT-LLM/tree/main/tests#1-how-does-the-ci-work))
- Any new dependencies have been scanned for license and vulnerabilities
- [CODEOWNERS](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/CODEOWNERS) updated if ownership changes
- Documentation updated as needed
- Update [tava architecture diagram](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/tava_architecture_diagram.md) if there is a significant design change in PR.
- The reviewers assigned automatically/manually are appropriate for the PR.

- [x] Please check this after reviewing the above items as appropriate for this PR.

## GitHub Bot Help

`/bot [-h] ['run', 'kill', 'skip', 'reuse-pipeline'] ...`

Provide a user friendly way for developers to interact with a Jenkins server.

Run `/bot [-h|--help]` to print this help message.

See details below for each supported subcommand.

<details>

`run  [--reuse-test (optional)pipeline-id --disable-fail-fast --skip-test --stage-list "A10-PyTorch-1, xxx" --gpu-type "A30, H100_PCIe" --test-backend "pytorch, cpp" --add-multi-gpu-test --only-multi-gpu-test --disable-multi-gpu-test --post-merge --extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx" --detailed-log --debug(experimental)]`

Launch build/test pipelines. All previously running jobs will be killed.

`--reuse-test (optional)pipeline-id ` *(OPTIONAL)* : Allow the new pipeline to reuse build artifacts and skip successful test stages from a specified pipeline or the last pipeline if no pipeline-id is indicated. If the Git commit ID has changed, this option will be always ignored. The DEFAULT behavior of the bot is to reuse build artifacts and successful test results from the last pipeline.

`--disable-reuse-test ` *(OPTIONAL)* : Explicitly prevent the pipeline from reusing build artifacts and skipping successful test stages from a previous pipeline. Ensure that all builds and tests are run regardless of previous successes.

`--disable-fail-fast ` *(OPTIONAL)* : Disable fail fast on build/tests/infra failures.

`--skip-test ` *(OPTIONAL)* : Skip all test stages, but still run build stages, package stages and sanity check stages. Note: Does **NOT** update GitHub check status.

`--stage-list "A10-PyTorch-1, xxx"` *(OPTIONAL)* : Only run the specified test stages. Examples: "A10-PyTorch-1, xxx". Note: Does **NOT** update GitHub check status.

`--gpu-type "A30, H100_PCIe"` *(OPTIONAL)* : Only run the test stages on the specified GPU types. Examples: "A30, H100_PCIe". Note: Does **NOT** update GitHub check status.

`--test-backend "pytorch, cpp"` *(OPTIONAL)* : Skip test stages which don't match the specified backends. Only support [pytorch, cpp, tensorrt, triton]. Examples: "pytorch, cpp" (does not run test stages with tensorrt or triton backend). Note: Does **NOT** update GitHub pipeline status.

`--only-multi-gpu-test ` *(OPTIONAL)* : Only run the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--disable-multi-gpu-test ` *(OPTIONAL)* : Disable the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--add-multi-gpu-test ` *(OPTIONAL)* : Force run the multi-GPU tests in addition to running L0 pre-merge pipeline.

`--post-merge ` *(OPTIONAL)* : Run the L0 post-merge pipeline instead of the ordinary L0 pre-merge pipeline.

`--extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx"` *(OPTIONAL)* : Run the ordinary L0 pre-merge pipeline and specified test stages. Examples: --extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx".

`--detailed-log ` *(OPTIONAL)* : Enable flushing out all logs to the Jenkins console. This will significantly increase the log volume and may slow down the job.

`--debug ` *(OPTIONAL)* : **Experimental feature**. Enable access to the CI container for debugging purpose. Note: Specify exactly one stage in the `stage-list` parameter to access the appropriate container environment. Note: Does **NOT** update GitHub check status.

For guidance on mapping tests to stage names, see `docs/source/reference/ci-overview.md`
and the `scripts/test_to_stage_mapping.py` helper.

### kill

`kill  `

Kill all running builds associated with pull request.

### skip

`skip --comment COMMENT `

Skip testing for latest commit on pull request. `--comment "Reason for skipping build/test"` is required. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

### reuse-pipeline

`reuse-pipeline `

Reuse a previous pipeline to validate current commit. This action will also kill all currently running builds associated with the pull request. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

</details>

## Summary by CodeRabbit

* **New Features**
  * Added vocabulary sharding support for guided decoding across tensor parallelism partitions via a new `rank` parameter.

* **Bug Fixes**
  * Enhanced bitmask validation to correctly handle sharded vocabulary configurations.

* **Refactor**
  * Refactored bitmask handling to use stride-based calculations instead of size-based, improving compatibility with non-contiguous memory layouts and sharded vocabularies.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>

### CODE REVIEW

File: tensorrt_llm/_torch/pyexecutor/guided_decoder.py [hunk 1]
```diff
@@ -145,11 +145,13 @@ def __init__(self,
                  guided_decoding_config: GuidedDecodingConfig,
                  max_num_sequences: int,
                  vocab_size_padded: int,
-                 max_num_draft_tokens: int = 0):
+                 max_num_draft_tokens: int = 0,
+                 rank: int = 0):
```
Comment (QiJune) [inline tensorrt_llm/_torch/pyexecutor/guided_decoder.py]: How about pass a `mapping` object instead of `rank`? So, we can do more safe check:


```
tp_size = mapping.tp_size
tp_rank = mapping.tp_rank
assert tp_size == vocab_size_padded // logits.size(1)
```

**Review by QiJune — APPROVED** [review]
LGTM
